op-plasma/commitment.go

```diff
@@ -16,12 +16,76 @@ var ErrCommitmentMismatch = errors.New("commitment mismatch")
 // CommitmentType is the commitment type prefix.
 type CommitmentType byte
 
-// KeccakCommitmentType is the default commitment type for the DA storage.
-const Keccak256CommitmentType CommitmentType = 0
+// CommitmentType describes the binary format of the commitment.
+// KeccakCommitmentType is the default commitment type for the centralized DA storage.
+// GenericCommitmentType indicates an opaque bytestring that the op-node never opens.
+const (
+	Keccak256CommitmentType CommitmentType = 0
+	GenericCommitmentType   CommitmentType = 1
+)
+
+// CommitmentData is the binary representation of a commitment.
+type CommitmentData interface {
+	Encode() []byte
+	TxData() []byte
+	Verify(input []byte) error
+}
 
-// Keccak256Commitment is the default commitment type for op-plasma.
+// Keccak256Commitment is an implementation of CommitmentData that uses Keccak256 as the commitment function.
 type Keccak256Commitment []byte
 
+// Keccak256Commitment is an implementation of CommitmentData that treats the commitment as an opaque bytestring.
+type GenericCommitment []byte
+
+// NewCommitmentData creates a new commitment from the given input and desired type.
+func NewCommitmentData(t CommitmentType, input []byte) CommitmentData {
+	switch t {
+	case Keccak256CommitmentType:
+		return NewKeccak256Commitment(input)
+	case GenericCommitmentType:
+		return NewGenericCommitment(input)
+	default:
+		return nil
+	}
+}
+
+// DecodeCommitmentData parses the commitment into a known commitment type.
+// The input type is determined by the first byte of the raw data.
+// The input type is discarded and the commitment is passed to the appropriate constructor.
+func DecodeCommitmentData(input []byte) (CommitmentData, error) {
+	if len(input) == 0 {
+		return nil, ErrInvalidCommitment
+	}
+	t := CommitmentType(input[0])
+	data := input[1:]
+	switch t {
+	case Keccak256CommitmentType:
+		return DecodeKeccak256(data)
+	case GenericCommitmentType:
+		return DecodeGenericCommitment(data)
+	default:
+		return nil, ErrInvalidCommitment
+	}
+}
+
+// NewKeccak256Commitment creates a new commitment from the given input.
+func NewKeccak256Commitment(input []byte) Keccak256Commitment {
+	return Keccak256Commitment(crypto.Keccak256(input))
+}
```

Comment:
I renamed `Keccak256` to `NewKeccak256Commitment` to stay consistent with other factory functions.